src/accountsdb/cache.zig

```diff
@@ -0,0 +1,228 @@
+pub const AccountsCache = struct {
+    lru: LRU,
+    allocator: std.mem.Allocator,
+
+    /// Atomically refcounted account
+    pub const CachedAccount = struct {
+        account: Account,
+        ref_count: ReferenceCounter,
+        /// the slot that this version of the account originates from
+        slot: Slot,
+
+        pub fn init(allocator: std.mem.Allocator, account: Account, slot: Slot) error{OutOfMemory}!CachedAccount {
+            return .{
+                .account = try account.clone(allocator),
+                .ref_count = .{},
+                .slot = slot,
+            };
+        }
+
+        pub fn deinit(self: *CachedAccount, allocator: std.mem.Allocator) void {
+            self.account.deinit(allocator);
+            self.* = undefined;
+        }
+
+        pub fn releaseOrDestroy(self: *CachedAccount, allocator: std.mem.Allocator) void {
+            if (self.ref_count.release()) {
+                self.deinit(allocator);
+                allocator.destroy(self);
+            }
+        }
+
+        pub fn copyRef(self: *CachedAccount) *CachedAccount {
+            if (self.ref_count.acquire()) {
+                return self;
+            } else {
+                unreachable;
+            }
+        }
+
+        fn getMeta(self: *CachedAccount) CachedAccountMeta {
+            return .{
+                .ref_count = self.ref_count,
+            };
+        }
+    };
+
+    /// A copy of a CachedAccout's fields, without the account
+    const CachedAccountMeta = struct {
+        ref_count: ReferenceCounter,
+    };
+
+    pub const LRU = LruCacheCustom(
+        .non_locking,
+        Pubkey,
+        *CachedAccount,
+        std.mem.Allocator,
+        CachedAccount.releaseOrDestroy,
+    );
+
+    const Self = @This();
+
+    pub fn init(
+        allocator: std.mem.Allocator,
+        max_items: usize,
+    ) error{OutOfMemory}!AccountsCache {
+        return .{
+            .lru = try LRU.initWithContext(allocator, max_items, allocator),
+            .allocator = allocator,
+        };
+    }
+
+    /// User is expected to release the returned CachedAccount.
+    pub fn get(self: *Self, pubkey: Pubkey, slot: Slot) ?*CachedAccount {
+        const account: *CachedAccount = self.lru.get(pubkey) orelse return null;
+        if (account.slot == slot) {
+            return account.copyRef();
+        } else {
+            return null;
+        }
+    }
+
+    /// Returns a copy of the CachedAccount's meta. Likely shouldn't be used outside of tests.
+    fn getMeta(self: *Self, pubkey: Pubkey, slot: Slot) ?CachedAccountMeta {
+        const account: *CachedAccount = self.lru.peek(pubkey) orelse return null;
+        if (account.slot == slot) {
+            return account.getMeta();
+        } else {
+            return null;
+        }
+    }
+
+    /// Does not influence underlying LRU's cache ordering
+    pub fn contains(self: *const Self, pubkey: Pubkey) bool {
```

Comment:
This function isn't actually used anywhere, plus users can just call `self.lru.contains` without much hassle.